File: code/binding/development/tools/cast/src/c++/cast/core/CASTComponent.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */
 
#include "CASTComponent.hpp"

using namespace std;

namespace cast {
  const std::string CASTComponent::END_COLOUR_ESCAPE = "\033[0m";

  CASTComponent::CASTComponent(const string &_id) : 
    InspectableComponent(_id),
    m_bLogOutput(false),
    m_bDebugOutput(false),
    m_startColourEscape(END_COLOUR_ESCAPE),
    m_startCalled(false),
    m_configureCalled(false) {
    m_pUnlockNotificationCondition
      = new omni_condition(&m_unlockNotificationMutex);


    static bool init = false;
    if(!init) {
      initCASTDatatypes();
      init = true;
    }

  }


  CASTComponent::~CASTComponent() {
    m_unlockNotificationMutex.lock();
    m_pUnlockNotificationCondition->broadcast();
    m_unlockNotificationMutex.unlock();
    delete m_pUnlockNotificationCondition;
  }

  void CASTComponent::stop() {
    //FrameworkProcess::
    InspectableComponent::stop();

    //release sleeping threads
    m_unlockNotificationMutex.lock();
    m_pUnlockNotificationCondition->broadcast();
    m_unlockNotificationMutex.unlock();
  }

  void CASTComponent::println(const char *format, ...) const {
    char msg[1024];
    va_list arg_list;
    va_start(arg_list, format);
    vsnprintf(msg, 1024, format, arg_list);
    va_end(arg_list);
    startColourise(cout);
    cout<<"[\""<<getProcessIdentifier()<<"\": "<< msg<<"]"; 
    endColourise(cout);
    cout<<endl;
  }

  void CASTComponent::println(const string & _s) const {
    println(_s.c_str());
  }


  void CASTComponent::log(const char *format, ...) const {
    if(m_bLogOutput) {
      char msg[1024];
      va_list arg_list;
      va_start(arg_list, format);
      vsnprintf(msg, 1024, format, arg_list);
      va_end(arg_list);
      startColourise(cout);
      cout<<"[LOG \""<<getProcessIdentifier()<<"\": "<<msg<<"]"; 
      endColourise(cout);
      cout<<endl;
    }  
  }

  void CASTComponent::log(const string & _s) const {
    log(_s.c_str());
  }




  void CASTComponent::debug(const char * format, ...) const {
    if(m_bDebugOutput) {
      
      char msg[1024];
      va_list arg_list;
      va_start(arg_list, format);
      vsnprintf(msg, 1024, format, arg_list);
      va_end(arg_list);
      startColourise(cout);
      cout<<"[DEBUG \""<<getProcessIdentifier()<<"\": "<< msg <<"]"; 
      endColourise(cout);
      cout<<endl;
    }
  }

  void CASTComponent::debug(const string & _s) const {
    debug(_s.c_str());
  }




  void* CASTComponent::run_undetached(void *arg) {

    //assert that start has been called by the subclass
    assert(m_startCalled);
    //assert that configure has been called by the subclass
    assert(m_configureCalled);

    try {

      runComponent();
      int* rv = new int(0);
      return (void*)rv;
    }
    catch(const BALTException & e) {
      println("Aborting after catching BALTException exception from runComponent()");
      println("what(): %s", e.what());
      std::abort();
    }
    catch(const std::exception &e) {
      println("WorkingMemoryChangeThread::forwardToSubclass std::exception caught");
      println(e.what());
      std::abort();
    }
    catch(const CORBA::BAD_PARAM &e) {
      println("Aborting after catching CORBA::BAD_PARAM exception from runComponent()");
      println("_name(): %s", e._name());
      std::abort();
    }
    catch(const CORBA::Exception &e) {
      println("Aborting after catching CORBA::Exception exception from runComponent()");
      println("_name(): %s", e._name());
      std::abort();
    }
    catch(...) {
      println("Aborting after catching unknown exception from runComponent()");
      std::abort();
    }



  }
  
  void CASTComponent::start() {
    InspectableComponent::start();
    m_startCalled = true;
  }

  void CASTComponent::configure(map<string,string> & _config) {

    //cout<<"CASTProcessingComponent::configure: "<<_config<<endl;
    
    m_configureCalled = true;

    map<string,string>::const_iterator i = _config.find(cdl::LOG_KEY);

    if(i != _config.end()) {
      string logValue = i->second;
      if(logValue == "true") {
	m_bLogOutput = true;
      }
      else if(logValue == "false") {
	m_bLogOutput = false;
      }
      else {
	println(string("config err, unknown value for log") + logValue); 
      }   
    }

    i = _config.find(cdl::DEBUG_KEY);

    if(i != _config.end()) {
      string logValue = i->second;
      if(logValue == "true") {
	m_bDebugOutput = true;
      }
      else if(logValue == "false") {
	m_bDebugOutput = false;
      }
      else {
	println(string("config err, unknown value for log") + logValue); 
      }   
    }

    i = _config.find(cdl::COMPONENT_NUMBER_KEY);

    assert(i != _config.end());
    string numberString = i->second;
    int myNumber = atoi(numberString.c_str());
    int printNumber = (myNumber % 7);
    int bold = (myNumber / 7) % 2;

    if(printNumber == 0) {
      //0 = do nothing
      m_startColourEscape = "\033[0m";
    }
    else {
      //otherwise use the connected colour
      ostringstream outStream;
      outStream<<"\033[3"<<printNumber;
      if(bold != 0) {
	outStream<<";1m";      
      }
      else {
	outStream<<"m";      
      }
      m_startColourEscape = outStream.str();
    }

    //println("MEMEMEMEMEMEMEME");
  }
  /**
   * Acquire the semaphore for access to this component.
   */

  std::ostream & 
  CASTComponent::startColourise(std::ostream &_stream) const {
    _stream<<m_startColourEscape;
    return _stream;
  }
  
  std::ostream & 
  CASTComponent::endColourise(std::ostream &_stream) const {
    _stream<<END_COLOUR_ESCAPE;
    return _stream;
  }


  void CASTComponent::lockProcess() {
    //println("locking");
    //m_semaphore.wait();
    m_mutex.lock();
    //println("locked");
  }

  /**
   * Release the semaphore for access to this component.
   */
  void CASTComponent::unlockProcess() {
    // println("unlocking");
    //m_semaphore.post();
    m_mutex.unlock();

    m_unlockNotificationMutex.lock();
    m_pUnlockNotificationCondition->broadcast();
    m_unlockNotificationMutex.unlock();  
    //  println("unlocked");

    omni_thread::yield();
    //self()->yield();

  }


  void CASTComponent::waitForUnlock() {
    m_unlockNotificationMutex.lock();
    m_pUnlockNotificationCondition->wait();
    m_unlockNotificationMutex.unlock();
  }

  //#include <errno.hpp>

  void CASTComponent::sleepProcess(unsigned long _millis) {
 
    int seconds = _millis / 1000;
    //  cout<<"secs: "<<seconds<<endl;
    _millis -= (seconds * 1000);
    //  int err = 0;
    long nanos = _millis * 1000000;
    //  cout<<"nanos: "<<nanos<<endl;
  
    omni_thread::sleep(seconds,nanos);

    //   timespec ts;
    //   ts.tv_sec = seconds;
    //   ts.tv_nsec = nanos;
    //   int err = nanosleep (&ts, NULL);

    //   if(err != 0) {
    //     throw CASTException(__HERE__, "failed to sleep: %s", strerror(err));
    //   }

  }



} //namespace cast 

File: code/binding/development/tools/cast/src/java/cast/core/CASTComponent.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * Basic components that are used to make up a CAAT architecture.
 */
package cast.core;

import java.util.Map;
import java.util.concurrent.Semaphore;

import Ice.Current;
import Ice.Identity;
import Ice.ObjectAdapter;
import cast.cdl.CASTTime;
import cast.cdl.COMPONENTNUMBERKEY;
import cast.cdl.DEBUGKEY;
import cast.cdl.LOGKEY;
import cast.interfaces.ComponentManagerPrx;
import cast.interfaces.TimeServerPrx;
import cast.interfaces._CASTComponentOperations;

/**
 * A basic component in the CAST framework. This class provides some basic
 * functionality useful for debugging
 * 
 * @author nah
 */
public abstract class CASTComponent implements _CASTComponentOperations,
		Runnable {

	/**
	 * The CAST time server.
	 */
	private TimeServerPrx m_timeServer;

	// determine whether debug output should be produced
	protected boolean m_bDebugOutput;

	// determine whether log output should be produced
	protected boolean m_bLogOutput;

	// access controller used to prevent asynchronous access
	protected Semaphore m_semaphore;

	protected boolean m_asleep;

	protected Object m_unlockNotification;

	private String m_startColourEscape;

	private boolean m_configureCalled;

	private boolean m_startCalled;

	private String m_componentID;

	private ComponentManagerPrx m_manager;

	private Thread m_runThread;

	// ice thingies
	private ObjectAdapter m_adapter;
	private Identity m_iceIdentity;

	/**
	 * Construct a new component with the given unique ID.
	 * 
	 * @param _id
	 *            The id used to identify this component.
	 */
	public CASTComponent() {
		// new fair (FIFO) semaphore with a single permit
		m_semaphore = new Semaphore(1, true);
		m_bLogOutput = false;
		m_bDebugOutput = false;
		m_unlockNotification = new Object();
		m_startCalled = false;
		m_configureCalled = false;
		m_startColourEscape = END_COLOUR_ESCAPE;
		m_componentID = null;
		m_manager = null;
	}

	protected CASTTime getCASTTime() {
		assert(m_timeServer != null);
		return m_timeServer.getCASTTime();
	}

	/**
	 * Print some debugging output.
	 * 
	 * @param _o
	 */
	protected void debug(Object _o) {
		String out = _o.toString();

		if (m_bDebugOutput) {
			String id = getComponentID().toString();

			// reduce overhead of string concat
			StringBuffer buff = new StringBuffer(12 + id.length()
					+ out.length());

			startColourise(buff);
			buff.append("[DEBUG \"");
			buff.append(id);
			buff.append("\": ");
			buff.append(out);
			buff.append("]");
			endColourise(buff);

			System.out.println(buff.toString());
		}
	}

	/**
	 * Use getComponentID instead.
	 * 
	 * @return
	 */
	@Deprecated
	public String getProcessIdentifier() {
		return getComponentID();
	}

	/**
	 * Gets the unique ID of this component.
	 * 
	 * @return
	 */
	public String getComponentID() {
		return m_componentID;
	}

	public String getID(Current __current) {
		return getComponentID();
	}

	/**
	 * Wait until the unlockComponent is successfully called.
	 */
	protected void waitForUnlock() {

		synchronized (m_unlockNotification) {
			try {
				m_unlockNotification.wait();
			} catch (InterruptedException e) {
				e.printStackTrace();
				System.exit(1);
			}
		}
	}

	/**
	 * Acquire the semaphore for access to this component.
	 */
	protected void lockComponent() {

		try {
			m_semaphore.acquire();
			// debug code to find deadlocks
			// if (!m_semaphore.tryAcquire(1, TimeUnit.SECONDS)) {
			// throw new RuntimeException("Unable to access queue!");
			// }
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Print some logging output.
	 * 
	 * @param _o
	 */
	protected void log(Object _o) {
		String out = _o.toString();

		if (m_bLogOutput) {

			String id = getComponentID();

			// reduce overhead of string concat
			StringBuffer buff = new StringBuffer(10 + id.length()
					+ out.length());
			startColourise(buff);
			buff.append("[LOG \"");
			buff.append(id);
			buff.append("\": ");
			buff.append(out);
			buff.append("]");
			endColourise(buff);
			System.out.println(buff.toString());

		}

	}

	/**
	 * Print out the input in a formatted way.
	 * 
	 * @param _s
	 *            String to print.
	 */
	protected void println(Object _o) {
		String out = _o.toString();
		String id = getComponentID();

		// reduce overhead of string concat
		StringBuffer buff = new StringBuffer(6 + id.length() + out.length());

		startColourise(buff);
		buff.append("[\"");
		buff.append(id);
		buff.append("\": ");
		buff.append(out);
		buff.append("]");
		endColourise(buff);

		System.out.println(buff.toString());

	}

	private void startColourise(StringBuffer _buff) {
		_buff.append(m_startColourEscape);
	}

	private void endColourise(StringBuffer _buff) {
		_buff.append(END_COLOUR_ESCAPE);
	}

	/**
	 * Method called in separate thread to run processing component. This method
	 * is called for each component when it is started by the framework process
	 * server.
	 */
	protected void runComponent() {
	}

	/**
	 * Put the processes thread to sleep for a number of milliseconds.
	 * 
	 * @param _millis
	 *            Number of milliseconds to sleep for.
	 */
	protected void sleepComponent(long _millis) {
		try {
			m_asleep = true;
			Thread.sleep(_millis);
			m_asleep = false;
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Release the semaphore for access to this component.
	 */
	protected void unlockComponent() {
		m_semaphore.release();
		synchronized (m_unlockNotification) {
			m_unlockNotification.notifyAll();
		}
	}

	private final static String END_COLOUR_ESCAPE = "\033[0m";

	protected void startInternal() {
		debug("CASTComponent.startInternal()");
		m_startCalled = true;
	}

	public final boolean isRunning() {
		return m_startCalled;
	}

	protected void start() {

	}

	protected void configure(Map<String, String> _config) {
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see
	 * balt.core.processes.FrameworkComponent#configure(java.util.Properties)
	 */
	protected void configureInternal(Map<String, String> __config) {
		// System.out.println("CAATComponent.configure()");

		m_configureCalled = true;

		String logString = __config.get(LOGKEY.value);

		if (logString != null) {
			m_bLogOutput = Boolean.parseBoolean(logString);
			// println("setting log to: " + m_bLogOutput);
		}

		String debugString = __config.get(DEBUGKEY.value);

		if (debugString != null) {
			m_bDebugOutput = Boolean.parseBoolean(debugString);
			// println("setting log to: " + m_bLogOutput);
		}

		String numberString = __config.get(COMPONENTNUMBERKEY.value);
		assert (numberString != null);
		int myNumber = Integer.parseInt(numberString);
		int printNumber = (myNumber % 7);
		int bold = (myNumber / 7) % 2;
		if (printNumber == 0) {
			// 0 = do nothing
			m_startColourEscape = "\033[0m";
		} else {
			// otherwise use the connected colour
			m_startColourEscape = "\033[3" + printNumber;
			if (bold != 0) {
				m_startColourEscape += ";1m";
			} else {
				m_startColourEscape += "m";
			}
		}
		// println("MEMEMEMEMEMEMEME");
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see balt.core.processes.FrameworkComponent#stop()
	 */
	protected void stopInternal() {
		try {
			m_runThread.join();
		} catch (InterruptedException e) {
			e.printStackTrace();
		}

		synchronized (m_unlockNotification) {
			m_unlockNotification.notifyAll();
		}
	}

	protected void stop() {
	}

	public void beat(Current __current) {
	}

	public void configure(Map<String, String> _config, Current __current) {
		configureInternal(_config);
		configure(_config);
	}

	public void run(Current __current) {
		m_runThread = new Thread(this);
		m_runThread.start();
	}

	public void setComponentManager(ComponentManagerPrx _man, Current __current) {
		m_manager = _man;
	}

	public void setID(String _id, Current __current) {
		assert (m_componentID == null);
		m_componentID = _id;
	}

	public void start(Current __current) {
		startInternal();
		start();
	}

	public void stop(Current __current) {
		m_startCalled = false;
		stop();
		stopInternal();
		// remove self from adapter
		// getObjectAdapter().remove(getIceIdentity());
	}

	/**
	 * Called when the thread starts at runtime
	 */
	public void run() {
		assert (m_startCalled);
		assert (m_configureCalled);
		try {
			runComponent();
		} catch (RuntimeException e) {
			println("Caught RuntimeException and rethrowing: " + e.getMessage());
			throw e;
		}
	}

	/**
	 * @param adapter
	 *            the adapter to set
	 */
	public void setObjectAdapter(ObjectAdapter adapter) {
		m_adapter = adapter;
	}

	/**
	 * @return the adapter
	 */
	public ObjectAdapter getObjectAdapter() {
		assert (m_adapter != null);
		return m_adapter;
	}

	/**
	 * @param iceIdentity
	 *            the iceIdentity to set
	 */
	public void setIceIdentity(Identity iceIdentity) {
		m_iceIdentity = iceIdentity;
	}

	/**
	 * @return the iceIdentity
	 */
	public Identity getIceIdentity() {
		assert (m_iceIdentity != null);
		return m_iceIdentity;
	}

	public void destroy(Current __current) {
		__current.adapter.remove(__current.id);
	}
	
	public void setTimeServer(TimeServerPrx _ts, Current __current) {
		m_timeServer = _ts;
	}
}
